### PR TITLE
Refactor/signup variant scripts

### PR DESCRIPTION
**Pull request by jumarcial**
# TLDR;
On the marketing side, We’re launching an A/B testing to iterate on the signup flow to reduce the 70% drop into the funnel. We’ve done 95% of the work on autonomy on the marketing side (design, UX, wording, javascript).**

We usually don’t ask engineers to review and approve the website code, but since the signup is particularly sensitive and is across the website and the product, we wanted to make sure we’re good before going live!

## Why this pull request
More context in this [Slack thread](https://gorgias.slack.com/archives/CD105FN21/p1686909004213379)
I had a call with @Yegor, who assigned @Mateusz Misiak to answer my questions about the account manager while I was re-writing the script, and then @Yegor Yakovishen assigned @maximelandry to review my code before going live.

I had a call with @maximelandry, who asked me to make a pull request to facilitate the code review, so here it is!

# What do I need?
 An engineer from the PE team to review the script we have written to confirm there is no unexpected issue or any security breach before releasing the experiment our A/B testing. 

**Expected deadline:** Next Tuesday (11/07/2023)
 
# Definition of done
signup-variant.js to be approved by PE engineers. We need to ensure their is no unexpected issue and that we have the same level a security as the current script from the account manager [here](https://accounts.gorgias.com/static/build/js/signup/signup.js)

## Details
src/js/signup-2.js, a duplicate of this script (src/js/signup-variant.js), is currently loaded in gorgias.com/signup-2 and gorgias.com/signup-2/account for testing purposes.
- It works as expected in terms of UX (already QA by our marketing brand designer and content marketing team).
- It also creates users an account as expected once the funnel is completed.

In case you need to load /src/js/signup-variant.js to gorgias.com/signup-2 or gorgias.com/signup-2/account for testing purposes, you can do it through the file src/main.js by replacing

`newScript(scriptBase + '/src/js/signup-2'+minBase+'.js','head',1);
`
with
`newScript(scriptBase + '/src/js/signup-variant'+minBase+'.js','head',1);
`
Find in the readme.txt of this repo the explanation to test the local code easily using a local server.

ps: I hope this pull request is clear enough (that's my first one). Don't hesitate to reach out if you have any questions

### CODE REVIEW

File: src/js/signup-variant.js [hunk 1]
```diff
@@ -1,225 +1,1046 @@
...
-              if (validationResult.reason === 'mailbox_does_not_exist') { ns.warning(self, 'Are you sure this email address exists?') }
+}
+
+// used to to fetch url parameters, set a session token key as local storage and prefilled fields using url parameters
+function processURLSearchParams () {
+
+  var searchParams = new URLSearchParams(window.location.search)
+  var sessionToken = searchParams.get(token_key)
+  var sessionEmail = searchParams.get(email_key.toLowerCase());
+  var sessionEmail = searchParams.get(email_key.toLowerCase());
```
Comment (maximelandry) [inline src/js/signup-variant.js:227]: Duplicated, overall let's use let instead over var

File: src/js/signup-variant.js [hunk 2]
```diff
@@ -1,225 +1,1046 @@
...
+      updateLinksForSession(sessionToken)
+  }
+  if(sessionEmail){
+    emailField.val(sessionEmail);
+  }
+  if(sessionfullname){
+    fullnameField.val(sessionfullname);
+  }
+  if(sessionPlanName){
+    window.localStorage.setItem(plan_name_key, sessionPlanName)
```
Comment (maximelandry) [inline src/js/signup-variant.js:242]: Why does it need to be stored? Is this used in different pages?

File: src/js/signup-variant.js [hunk 3]
```diff
@@ -1,225 +1,1046 @@
...
+  }
+  if(sessionPlanPeriod){
+    window.localStorage.setItem(plan_period_key, sessionPlanPeriod)
```
Comment (maximelandry) [inline src/js/signup-variant.js:245]: Why does it need to be stored? Is this used in different pages?

File: src/js/signup-variant.js [hunk 4]
```diff
@@ -1,225 +1,1046 @@
-(function (ns) {
-  let waitForAnalyticsIntervalCount = 0
-  let waitForAnalyticsIntervalMax = 3
-  let waitForAnalyticsInterval = null
-  ns.waitForAnalytics = function (callback) {
-    waitForAnalyticsInterval = window.setInterval(function () {
+const email_key = 'email';
+const fullname_key = 'fullname';
+const plan_name_key = 'plan_name';
+const plan_period_key = 'period';
```
Comment (maximelandry) [inline src/js/signup-variant.js:4]: Are these consts, local storage variables or dynamically used variables?

File: src/js/signup-variant.js [hunk 5]
```diff
@@ -1,225 +1,1046 @@
...
+    window.localStorage.setItem(plan_name_key, sessionPlanName)
+  }
+  if(sessionPlanPeriod){
+    window.localStorage.setItem(plan_period_key, sessionPlanPeriod)
+  }
+
+}
+
+// used ot check is email adress is personnal or not. Currently not used in our emailVerify function
+function isPersonalEmail(email) {
```
Comment (maximelandry) [inline src/js/signup-variant.js:251]: This function is never called and doing checks only in the frontend is not great + the list of domains in not complete

File: src/js/signup-variant.js [hunk 6]
```diff
@@ -1,225 +1,1046 @@
...
+        if (typeof window.GORGIAS_INIT_CALLBACK === 'function') {
+            window.GORGIAS_INIT_CALLBACK(data);
+        }
+      },
+      //error
+      function (response) {
+      },
+      //complete
+      function (){
+      }
```
Comment (maximelandry) [inline src/js/signup-variant.js:709]: could be removed?

File: src/js/signup-variant.js [hunk 7]
```diff
@@ -1,225 +1,1046 @@
...
+function signupAccountFormHandler(form){
+  var API_INIT_ENDPOINT = '/account/init';
+
+  post(
+    API_INIT_ENDPOINT,
+    null,
+    //success
+    function (data) {
+      if (data.redirect_url) {
+        window.location = '/signup-2'
```
Comment (maximelandry) [inline src/js/signup-variant.js:723]: Should probably be a const as the name will change

File: src/js/signup-variant.js [hunk 8]
```diff
@@ -1,225 +1,1046 @@
...
+      accountDomainVerify(status,companyDomainField.val(),false);
+      handleFormStatus (form,status,messaging);
+
+      return false; 
+    }
+
+    else{
+      status = 'valid';
+      window.localStorage.removeItem(account_domain_key + '-status');
+      window.localStorage.removeItem(company_domain_key + '-status');
```
Comment (maximelandry) [inline src/js/signup-variant.js:1025]: Same here, I'm not sure why local storage is used to store these

**Review by maximelandry — COMMENTED** [review]
I didn't see significant "security" changes but code could be clean up and trimmed down. Also I'm not sure about the benefits of the locale storage here

File: src/js/signup-variant.js [hunk 9]
```diff
@@ -1,225 +1,1046 @@
...
+var accountDomainTextInfoWrapper = $('#account-domain-info-wrapper') || "";
+var companyDomainField = $('#signup-account-form input[name="' + company_domain_key + '"]') || "";
+var accountDomainField = $('#signup-account-form input[name="' + account_domain_key + '"]') || "";
+var signupButton = $('input[id="signup-button"]') || "";
+var signupButtonLoading = $('#signup-button-loading') || "";
+var ssoGoogleButton = $('a[id="sso-button-google"]') || "";
+var userForm = $('form#signup-user-form');
+var accountForm = $('form#signup-account-form');
+var accountFormWrapper = $('#signup-account-form-wrapper');
+var accountFormLoadingWrapper = $('div#signup-account-form-loading-wrapper');
```
[the remark this answers is not in the document]
Comment (jumarcial) [reply src/js/signup-variant.js:29]: We use each variable in multiple functions, so it would be more efficient to declare them once rather than writing multiple times the $(anchor). It also reduces the risk of errors. But it's maybe a bad practice; let me know if it requires change.

File: src/js/signup-variant.js [hunk 10]
```diff
@@ -1,225 +1,1046 @@
...
+    //complete
+    function (){
+    }
+  )
+}
+
+// once submission start - clean all alert messaging and customize the button (disable + update value) an disable all fields to avoid update
+function onSubmitStart(form, button) {
+  form.find('input').prop("disabled", true);
+  button.addClass('hidden').css('dislplay','none');
```
Comment (Spark-NF) [inline src/js/signup-variant.js:751]: Typo on "dislplay". But Isn't `.hidden` already defined as `display:none`?
Also why not just use jQuery's `.hide()` / `.show()`?

File: src/js/signup-variant.js [hunk 11]
```diff
@@ -1,225 +1,1046 @@
...
+  });
+  
+  // checking password field without delay right after typing
+  passwordField.on('input', function(){
+    passwordVerify('warning');
+  })
+  
+  // used to display / hide the password field content using the button within the field
+  $('#transformButton').on('click', function() {
+    currentType = $('#signup-user-form input[name="' + password_key+ '"]').attr('type');
```
Comment (Spark-NF) [inline src/js/signup-variant.js:935]: NIT: you could introduce a variable for this `$('#signup-user-form input[name="' + password_key+ '"]')` to prevent repeating it thrice.

**Review by Spark-NF — CHANGES_REQUESTED** [review]

File: src/js/signup-variant.js [hunk 12]
```diff
@@ -1,225 +1,1046 @@
...
+    var accountSubdomainsApproved = JSON.parse(localStorage.getItem('account-subdomains-approved'));
+  } catch (error) {
+    var accountSubdomainsApproved = {}
+    localStorage.removeItem('account-subdomains-approved');
+  }
+
+
+  
+  if (localStorage.getItem('account-subdomains-approved') === null || !accountSubdomainsApproved.hasOwnProperty(account_domain_submitted) ) {
+    var API_BASE_URL = "https://us-central1-gorgias-growth-production.cloudfunctions.net/check_helpdesk_domain";
```
[the remark this answers is not in the document]
Comment (Spark-NF) [reply src/js/signup-variant.js:391]: Discussed with Max, this behaves differently and returns a valid domain if none is found. However, FYI doing HTTP calls and checking for 404 errors is not a reliable way to do that if for example a domain was just created, as the DNS might not have propagated yet. The source of truth is still in the account manager DB.

File: src/js/signup-variant.js [hunk 13]
```diff
@@ -1,225 +1,1030 @@
...
-      if (navigator.userAgent === DATADOG_BOT_USER_AGENT) { onSubmitUserSignupForm() }
-      else { grecaptcha.execute() }
+    .done(function (response) {
+      APIresponse = {
+        domain_exist:false,
+        account_domain: response.account_domain
+      }      
+      // Check if "account-subdomains-approved" exists in local storage
+      var data = localStorage.getItem('account-subdomains-approved') || {};
+      JSON.parse(data);
```
Comment (Spark-NF) [inline src/js/signup-variant.js:425]: Note: `JSON.parse({})` will raise an exception. Might want to use `"{}"` as default value instead, my bad.

Comment (jumarcial) [reply src/js/signup-variant.js:425]: Sorry I don't understand your feedback, what do you mean?

File: src/js/signup-variant.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (jumarcial) [reply src/js/signup-variant.js:391]: > However, FYI doing HTTP calls and checking for 404 errors is not a reliable way to do that if for example a domain was just created, as the DNS might not have propagated yet
I did the the following test multiple times
1. creating a new account with a new Gorgias subdomain
2. then immediately tried to create another account with the same Gorgias subdomain
And the cloud function returns the expected result: the gorgias subdomain sent, incremented by 1

Are you sure the DNS propagation matter? {{domain}}.gorgias.com/login page is created immediately once the account is created. Otherwise, it means a new user may not be able to visit their helpdesk login page immediately after creating their account, which is weird

[the remark this answers is not in the document]
Comment (jumarcial) [reply src/js/signup-variant.js:391]: We talking about a super rare case. Once again, it's an A/B test and if the result is positive, then I guess we'll have to migrate the cloud function logic into the account manager.
Furthermore, it's just the frontend check, meaning in any case if we submit the form to the account manager with an existing domain - the account-manager will return an error message that we display.

**Review by jumarcial — COMMENTED** [review]
Hey @maximelandry @Spark-NF - I followed up on all of your comments a couple of days ago, but they stayed in "Pending". Didn't know I had to submit a review to publish my 23 comments😅

File: src/js/signup-variant.js [hunk 14]
```diff
@@ -1,225 +1,1046 @@
...
-  ns.waitForAnalytics = function (callback) {
-    waitForAnalyticsInterval = window.setInterval(function () {
+const email_key = 'email';
+const fullname_key = 'fullname';
+const plan_name_key = 'plan_name';
+const plan_period_key = 'period';
+const password_key = 'password';
+const company_domain_key = 'company_domain';
+const account_domain_key = 'account_domain';
+var emailField = $('#signup-user-form input[name="'+ email_key + '"]') || "";
```
[the remark this answers is not in the document]
Comment (Spark-NF) [reply src/js/signup-variant.js:20]: It's not very important, but usually if something is "missing" it's better to use `null` to convey the meaning, since it's the symbol of a "missing" value. While an empty string could be other things. At the end of the day they're both "falsy" so in your case it doesn't really change anything.

File: src/js/signup-variant.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (Spark-NF) [reply src/js/signup-variant.js:391]: I was just pointing it out. I know it behaves differently and in a more convenient manner, but at the end of the day the source of truth for accounts is in the DB, not in the DNS. So if it's what is wanted long term, it should be moved to the account manager 😄

File: src/js/signup-variant.js [hunk 15]
```diff
@@ -1,225 +1,1046 @@
...
+    post(
+      API_VALIDATION_ENDPOINT,
+      validationData,
+      // success callback
+      function (data) {
+        if (data && data['signup-email-validation']) {
+          var validationResult = data['signup-email-validation']
+          if (validationResult.risk === 'high') {
+            if (validationResult.reason === 'mailbox_does_not_exist' || validationResult.result === "undeliverable") {
+              verifyMessage = 'Are you sure this email address exists?';
```
[the remark this answers is not in the document]
Comment (Spark-NF) [reply src/js/signup-variant.js:296]: This verification flag and messages are pretty confusing 😅

File: src/js/signup-variant.js (hunk 13 above)
Comment (Spark-NF) [reply src/js/signup-variant.js:425]: If the entry does not exist in the localStorage, `data` will have the value of `{}`.
Then, you will do `JSON.parse(data)` (also, shouldn't you put the result of this in some variables?), which is the equivalent to `JSON.parse({})` (since `data = {}`).

And it will fail:
image

That's because `JSON.parse` only accepts strings, i.e. `"{}"` instead of `{}`.

Comment (jumarcial) [reply src/js/signup-variant.js:425]: How ok! Thanks for flagging this. Is adding a "Try" makes more sense? 

 ```
     // Check if "account-subdomains-approved" exists in local storage
      var data = localStorage.getItem('account-subdomains-approved') || {};
      var dataObject;

      try {
        dataObject = JSON.parse(data);

      } catch (error) {
        dataObject = {};
      }
      
      dataObject[account_domain_submitted] = APIresponse.account_domain;
      localStorage.setItem('account-subdomains-approved', JSON.stringify(dataObject));
```

File: src/js/signup-variant.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (jumarcial) [reply src/js/signup-variant.js:20]: ok good to know! thank you for the lesson 🤓

File: src/js/signup-variant.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (jumarcial) [reply src/js/signup-variant.js:391]: We're aligned- if XP is successful I'll make a request through PMs to include this into the roadmap for more stable implementation 👍

File: src/js/signup-variant.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (jumarcial) [reply src/js/signup-variant.js:296]: The VerifyStatus parameter is always set as if it didn't pass the validation.

File: src/js/signup-variant.js (hunk 6 above)
Comment (maximelandry) [reply src/js/signup-variant.js:711]: I think it should yes. I think you and remove and test to see

File: src/js/signup-variant.js (hunk 8 above)
Comment (maximelandry) [reply src/js/signup-variant.js:1018]: Not necessarily I was curious about the flow and why we needed them to be stored in the first place

**Review by maximelandry — APPROVED** [review]
I won't block this one as it's an experiment and you own it. Didn't test it on my end and comments were mostly about the code than what it does. If @Spark-NF is ok, I'm ok too.

File: src/js/signup-variant.js (hunk 13 above)
Comment (Spark-NF) [reply src/js/signup-variant.js]: That works, but you could have also just used a `|| "{}"` instead of `|| {}` on the localStorage line, since `JSON.parse("{}")` would work and return `{}` (or even do both that and a try/catch).

**Review by Spark-NF — APPROVED** [review]
